### ISSUE TITLE
How to notify clients outside of a hub context (documented way does not work)

### ISSUE THREAD

**Issue by WernMich** [aspnet/SignalR#2935]
Hi,

I am trying to notify my clients as documented on several places (e.g. https://docs.microsoft.com/de-de/aspnet/core/signalr/hubcontext?view=aspnetcore-2.1 topic: "Get an instance of IHubContext in middleware") outside of a hub object (no previous call from a client). There is no website within the actual project so there is no "Controller" class where IHubContext can be injected via DI.

How do I get an IHubContext object to notify my clients? The documented version does not even compile!

Startup.cs class, method "public void Configure(IApplicationBuilder app, IHostingEnvironment env)":

app.Use(next => (context) =>
            {
                var hubContext = (IHubContext<NotificationHub>)context
                    .RequestServices
                    .GetServices<IHubContext<NotificationHub>>();
                //...
            });

causes:

Startup.cs(58,39,58,41): error CS1643: Not all code paths return a value in lambda expression of type 'RequestDelegate'
Startup.cs(58,29,58,41): error CS1662: Cannot convert lambda expression to intended delegate type because some of the return types in the block are not implicitly convertible to the delegate return type

The delegate type has a completely different layout: returning Task and consuming HttpContext object.
So I tried to change the signature to this resulting code (I am using two hubs):

app.UseSignalR(routes =>
            {
                routes.MapHub<CommandOutputHub>("/CommandOutputHub");
                routes.MapHub<NotificationHub>("/NotificationHub");
            });

app.Use(next => (context) =>
            {
                NotificationHubContext = (IHubContext<NotificationHub>)context
                    .RequestServices
                    .GetServices<IHubContext<NotificationHub>>().First();

                return Task.CompletedTask;
                //...
            });

But the lambda becomes never called. Even when I exchange the "null" task with a valid one, this task becomes never executed.

Can anyone provide a complete sample, how notification of clients can be done from outside a Hub object without using a Controller class?

Server-Side: 
Microsoft.AspNetCore.App 2.1.1 incl. Microsoft.AspNetCore.SignalR (1.0.1) (.NET console app or service hosted on IIS Express, C# .NET Core 2.1, Win 10 Pro OS)

Client:
Microsoft.AspNetCore.SignalR.Client 1.0.3 (.NET console app, C# .NET Core 2.1, Win 10 Pro OS)

**Comment by anurse** [comment]
There's a sample referenced right at the top of the doc you linked: https://github.com/aspnet/Docs/tree/master/aspnetcore/signalr/hubcontext/sample/

**Comment by anurse** [comment]
> Startup.cs(58,39,58,41): error CS1643: Not all code paths return a value in lambda expression of type 'RequestDelegate'
> Startup.cs(58,29,58,41): error CS1662: Cannot convert lambda expression to intended delegate type because some of the return types in the block are not implicitly convertible to the delegate return type

These are compilation errors, likely because you didn't use the `async` keyword in the delegate. The docs are missing that though, so I submitted https://github.com/aspnet/Docs/pull/8480 to fix that. Sorry about that!

**Comment by WernMich** [comment]
Hi, thank you for your fast response! The sample you referenced is not what I am looking for (as described, I do not have a website on this "site", so there is no Controller here anywhere!). And as I already mentioned too, I got the code compiling but the lambda (RequestDelegate) is never called. I need a working notification from "middleware" code: clients simply connect to the "NotificationHub" and become notified after a certain time from the server - how to achieve this? A sample would be really awesome!

**Comment by WernMich** [comment]
Hi again, I found a working solution on Stack Overflow(https://stackoverflow.com/a/38832879): 

Adding a IServiceProvider object to the argument list of the Configure method (for me still magic this auto DI). Then calling GetService<IHubContext<NotificationHub>>() on this object returns the HubContext reference I needed. If there is a more fancy way to do this then please add it to the documentation too else maybe provide this working solution.

Best,
Michael

**Comment by anurse** [comment]
Yes, you need to get `IHubContext` out of DI. So if you don't use Controllers, then you need to access it in some other way. If you can provide a more detailed explanation of how you need to access the IHubContext we can look at trying to provide better documentation here.

> I need a working notification from "middleware" code

If the above code wasn't working for middleware, can you post a complete sample that reproduces the error? That's definitely a scenario we've tested. Middleware should be called for each incoming request, so if your registered middleware isn't being called it could be because a middleware higher up the chain is being called.

>     NotificationHubContext = (IHubContext<NotificationHub>)context
>         .RequestServices
>         .GetServices<IHubContext<NotificationHub>>().First();

This code is very odd. You're taking the HubContext from a single request and storing it somewhere. [Middleware is called for each individual request](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/?view=aspnetcore-2.1), so I'm not sure I understand what you're trying to achieve. If you want to trigger notifications from a background service, you not be using middleware for that as middleware is for request handling, not for background services. If you are using a background service you should use an [`IHostedService`](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/hosted-services?view=aspnetcore-2.1) which is DI-activated, so you can have `IHubContext<NotificationHub>` in your constructor and use it freely there.
